src/main/java/com/autonomic/tmc/auth/BaseClientCredentialsTokenSupplier.java

```diff
@@ -0,0 +1,203 @@
+        this.clientId = clientId;
+        this.clientSecret = clientSecret;
+        //TODO: Add validation for clientId and Secret
+        this.tokenUrl = (tokenUrl != null) ? tokenUrl : DEFAULT_TOKEN_URL;
+        try {
+            this.tokenEndpoint = new URI(this.tokenUrl);
+        } catch (URISyntaxException e) {
+            throw new SdkClientException(
+                String.format("tokenUrl [%s] is not a valid URL", tokenUrl), e);
+        }
+    }
+
+    /**
+     * This method responds with a valid String representation of the Bearer token.  If a token
+     * becomes expired, a new valid token will be requested automatically. This method will always
+     * return a valid token.
+     *
+     * @return String representation of Bearer token.
+     * @throws AuthenticationCommunicationException Thrown when an unexpected condition is
+     *                                              encountered while making the client credentials
+     *                                              grant POST
+     * @throws AuthenticationFailedException        Thrown when the credentials that were provided
+     *                                              are expressly rejected.
+     */
+    @Override
+    public synchronized String get() {
+        if (token != null && !token.isExpired()) {
+            log.debug("Cached token found.");
+            return token.getValue();
+        }
+        resetToken();
+        return token.getValue();
+    }
+
+    private synchronized void resetToken() {
+        // if not already reset
+        if (token == null || token.isExpired()) {
+            log.debug("No cached token. Retrieving a new token.");
+            token = authenticate();
+        }
+    }
+
+    private Token authenticate() {
+
+        TokenResponse response = executeTokenRequest();
+        AccessToken accessToken = processTokenResponse(response);
+
+        return buildToken(accessToken);
+    }
+
+    private TokenRequest createTokenRequest() {
+        AuthorizationGrant clientGrant = new ClientCredentialsGrant();
+        ClientID clientID = new ClientID(clientId);
+        Secret secret = new Secret(this.clientSecret);
+        ClientAuthentication clientAuthentication = new ClientSecretPost(clientID, secret);
+
+        return new TokenRequest(this.tokenEndpoint, clientAuthentication, clientGrant, null);
+    }
+
+    AccessToken processTokenResponse(TokenResponse response) {
+
+        if (!response.indicatesSuccess()) {
+            HTTPResponse httpResponse = response.toHTTPResponse();
+
+            int responseCode = httpResponse.getStatusCode();
+            if (responseCode == 401 || responseCode == 400) {
+                //TODO: pass in the error code object
+                throw new SdkServiceException(String
+                    .format("Authorization failed for user [%s] at tokenUrl [%s]",
+                        this.clientId, this.tokenUrl), responseCode);
```

Comment:
Modify this Exception so we can pass the TokenErrorResponse in here instead.

Then get the statusCode inside the exception.

```
TokenErrorResponse tokenErrorResponse = response.toErrorResponse();
```